torch/_dynamo/symbolic_convert.py

```diff
@@ -1166,8 +1171,50 @@ def POP_FINALLY(self, inst):
         if preserve_tos:
             self.push(tos)  # type: ignore[possibly-undefined]
 
-    def FOR_ITER(self, inst):
+    def FOR_ITER(self, inst: Instruction):
         it = self.pop().realize()
+        if (
+            isinstance(it, variables.RangeIteratorVariable)
+            and inst.starts_line is not None
+        ):
+            try:
+                # Converts a loop to a function body, to benefit
+                # from function compilation caching.
+                assert inst.target is not None
+                op = RangeHigherOrderVariable.make_self(
+                    self,
+                    it,
+                    self.f_globals,
+                    self.f_code,
+                    self.instructions[
+                        self.instruction_pointer : self.indexof[inst.target]
+                    ],
+                    self.symbolic_locals,
+                )
+
+                new_locals = op.functionalize(self)
+                for loc in new_locals:
+                    if isinstance(loc, torch.fx.Proxy):
+                        del loc.node.meta["example_value"]
+                args = [
+                    wrap_fx_proxy(self, loc) if isinstance(loc, torch.fx.Proxy) else loc
+                    for loc in new_locals
+                ]
+                for name, v in zip(self.f_code.co_varnames, args):
+                    if (
+                        isinstance(v, ConstantVariable)
+                        and v.as_python_constant()
+                        is RangeHigherOrderVariable.NOT_SET_SENTINEL
+                    ):
+                        continue
+                    self.symbolic_locals[name] = v
```

Comment:
what happens if the body of the loop adds more locals on each iteration? Do we graph break successfully on that?